python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+                    nSig = self.config.nSigmaBright if imageType == 'flat' else self.config.nSigmaDark
+                    self._plot(dataRef, exp, visit, nSig, defects, imageType)
+
+        self.log.info(f'Combining {len(defectListFromDarks)} defect sets from darks for detector {detNum}')
+        mergedDefectsFromDarks = self._postProcessDefectSets(defectListFromDarks, exp.getDimensions())
+        self.log.info(f'Combining {len(defectListFromFlats)} defect sets from flats for detector {detNum}')
+        mergedDefectsFromFlats = self._postProcessDefectSets(defectListFromFlats, exp.getDimensions())
+
+        self.log.info(f'Combining bright and dark defect sets')
+        brightDarkPostMerge = [mergedDefectsFromDarks, mergedDefectsFromFlats]
+        allDefects = self._postProcessDefectSets(brightDarkPostMerge, exp.getDimensions(), mode='OR')
+
+        self._writeData(dataRef, allDefects)
+
+        self.log.info('Finished finding defects in detector %s' % detNum)
+        return pipeBase.Struct(exitStatus=0)
+
+    @staticmethod
+    def _nPixFromDefects(defects):
+        """TODO: Write docstring
+        """
+        nPix = 0
+        for defect in defects:
+            nPix += defect.getBBox().getArea()
+        return nPix
+
+    def _writeData(self, dataRef, defects):
+        """Write the data out to the defect file.
+
+        TODO: Add parameters
+        """
+        filename = dataRef.getUri(write=True)  # does not guarantee that full path exists
+        dirname = os.path.dirname(filename)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        self.log.info(f'Writing defects to {filename[:-5]} in format: {self.config.writeAs}')
+
+        if self.config.writeAs in ['FITS', 'BOTH']:
+            defects.writeFits(filename)
+        if self.config.writeAs in ['ASCII', 'BOTH']:
+            defects.writeText(filename)
+        return
+
+    @staticmethod
+    def _getRunListFromVisits(butler, visitList):
+        """Return the set of runs for the visits in visitList."""
+        runs = set()
+        for visit in visitList:
+            runs.add(butler.queryMetadata('raw', 'run', dataId={'visit': visit})[0])
+        return runs
+
+    def _postProcessDefectSets(self, defectList, imageDimensions, mode=None):
+        """Combine a list of defects to make a single defect object.
+
+        AND, OR or use percentage of visits in which defects appear
+        depending on config.
+
+        Parameters
+        ----------
+        defectList: `list` of `lsst.meas.algorithms.interp.Defect`
+            List of defects to merge
+
+        imageDimensions: `tuple` of `int`
+            The size of the image
+
+        Returns
+        -------
+        defects: `lsst.meas.algorithms.interp.Defect`
+            The merged defects
+        """
+        # so that empty lists can be passed in for input data
+        # where only flats or darks are supplied
+        if not defectList:
```

Comment:
I guess I wasn't thinking about what it would be, and so used that to cover the last two at the same time. I think it should always be an empty list though, so I'll change it to `if defectList == []`